python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py

```diff
@@ -0,0 +1,1113 @@
+        l2_idx, l1_idx, r1_idx = boresch.restraint.angles[0].atomgroup.atoms.ix
+        l1_idx, r1_idx, r2_idx = boresch.restraint.angles[1].atomgroup.atoms.ix
+        l3_idx, l2_idx, l1_idx, r1_idx = boresch.restraint.dihedrals[
+            0].atomgroup.atoms.ix
+        l2_idx, l1_idx, r1_idx, r2_idx = boresch.restraint.dihedrals[
+            1].atomgroup.atoms.ix
+        l1_idx, r1_idx, r2_idx, r3_idx = boresch.restraint.dihedrals[
+            2].atomgroup.atoms.ix
+
+        # The index of the best frame
+        index = boresch.restraint.min_frame
+        # r1-l1 (r0, kr)
+        r0 = boresch.restraint.bond.values[index] * _angstrom
+        kr = 10 * _kcal_per_mol / (_angstrom ** 2)
+        # r2-r1-l1 (thetaA0, kthetaA)
+        thetaA0 = boresch.restraint.angles[1].values[index] * _degree
+        kthetaA = 10 * _kcal_per_mol / (_radian ** 2)
+        # r1-l1-l2 (thetaB0, kthetaB)
+        thetaB0 = boresch.restraint.angles[0].values[index] * _degree
+        kthetaB = 10 * _kcal_per_mol / (_radian ** 2)
+        # r3-r2-r1-l1 (phiA0, kphiA)
+        phiA0 = boresch.restraint.dihedrals[2].values[index] * _degree
+        kphiA = 10 * _kcal_per_mol / (_radian ** 2)
+        # r2-r1-l1-l2 (phiB0, kphiB)
+        phiB0 = boresch.restraint.dihedrals[1].values[index] * _degree
+        kphiB = 10 * _kcal_per_mol / (_radian ** 2)
+        # r1-l1-l2-l3 (phiC0, kphiC)
+        phiC0 = boresch.restraint.dihedrals[0].values[index] * _degree
+        kphiC = 10 * _kcal_per_mol / (_radian ** 2)
+
+        restraint_dict = {
+            # The default index is in the format of numpy.int64
+            # So need to convert to int
+            "anchor_points": {"r1": system.getAtom(int(r1_idx)),
+                              "r2": system.getAtom(int(r2_idx)),
+                              "r3": system.getAtom(int(r3_idx)),
+                              "l1": system.getAtom(int(l1_idx)),
+                              "l2": system.getAtom(int(l2_idx)),
+                              "l3": system.getAtom(int(l3_idx))},
+            "equilibrium_values": {"r0": r0,
+                                   "thetaA0": thetaA0,
+                                   "thetaB0": thetaB0,
+                                   "phiA0": phiA0,
+                                   "phiB0": phiB0,
+                                   "phiC0": phiC0},
+            "force_constants": {"kr": kr,
+                                "kthetaA": kthetaA,
+                                "kthetaB": kthetaB,
+                                "kphiA": kphiA,
+                                "kphiB": kphiB,
+                                "kphiC": kphiC
+                                }}
+        # TODO: extract the best frame and feed it into Restraint
+        # Waiting for the BSS to fix the getFrames
+        # best_frame = traj.getFrames(index)
+        best_frame = system
+        restraint = Restraint(best_frame, restraint_dict,
```

Comment:
So the problem is how to pass the restraint information.
My current plan is that the user would generate the `Restraint` using `RestraintSearch` and then pass the information to process. I don't think user is able to generate the restraint by themselves as unless the restraint are carefully generated, it will almost always explode.

Yes, I can make Restraint into a private class.

Reply:
Great, that was exactly what I was thinking :+1:

Reply:
That's a good point. I guess a dictionary for the parameters makes sense here since it's a single object that could be pickled easily.